protocol/decoupled-hardfork-features.md

```diff
@@ -0,0 +1,308 @@
+  OptimismMintableERC20FactoryProxy = "0x75505a97BD334E7BD3C476893285569C4136Fa0F"
+  OptimismPortalProxy = "0xbEb5Fc579115071764c7423A4f12eDde41f106Ed"
+  SystemConfigProxy = "0x229047fed2591dbec1eF1118d64F7aF3dB9EB290"
+  ProxyAdmin = "0x543bA4AADBAb8f9025686Bd03993043599c6fB04"
+  AnchorStateRegistryProxy = "0x1c68ECfbf9C8B1E6C0677965b3B9Ecf9A104305b"
+  DelayedWETHProxy = "0x21429aF66058BC3e4aE4a8f2EC4531AaC433ecbC"
+  DisputeGameFactoryProxy = "0xe5965Ab5962eDc7477C8520243A95517CD252fA9"
+  FaultDisputeGame = "0x5738a876359b48A65d35482C93B43e2c1147B32B"
+  MIPS = "0xF027F4A985560fb13324e943edf55ad6F1d15Dc1"
+  PermissionedDisputeGame = "0x1Ae178eBFEECd51709432EA5f37845Da0414EdFe"
+  PreimageOracle = "0x1fb8cdFc6831fc866Ed9C51aF8817Da5c287aDD3"
+```
+
+# Alternatives Considered
+
+### **Individual Feature Timestamps**
+
+While the feature-to-hardfork mapping adds flexibility within hardfork feature sets,
+a more complete decoupling could use per-feature activation timestamps directly in
+`RollupConfig`. For example, a `map[string]uint64` where key is feature name and value
+is its independent timestamp. Hardforks would then serve as non-binding labels or presets
+(e.g., "isthmus" predefined as activating features X, Y, Z at time T), loaded optionally
+for superchain coordination.
+
+> [!NOTE]
+>
+> In practice features are never actually “bound” to a specific hardfork at the clien
+> level since they can be disabled or enabled via cli flags or a custom rollup config.
+
+This approach:
+
+- Eliminates hardfork dependency entirely for feature checks.
+- Simplifies overrides (direct timestamp per feature via CLI).
+- Enhances modularity for custom chains or phased rollouts.
+- Risks less coordination if not enforced via governance/tools.
+
+It simplifies the level of indirection between hardfork and individual features, as it
+avoids indirect lookups while still supporting bundled activations through presets.
+
+On the other hand, it increases the surface area for misconfigurations where timestamps
+are incorrectly set for various features. The surface area is increased since there are
+many more features than hardforks. If one client uses the backwards-compatible method of
+implementing the feature set, and another doesn’t, one wrong timestamp could result in a
+chain split.
+
+### **Leave it Unchanged**
+
+One option (a bit of a steel man argument) is to leave the current method unchanged.
+One could argue that the problem statement is a result of the hardfork upgrade process
+- not the actual implementation method. Instead of handling this at the implementation
+level, fix the upgrade process so that a hardfork’s feature set is properly established
+prior to implementation.
+
+# Risks
+
+- **Increased Complexity:** Adding maps/methods complicates config management and client
+  logic, risking bugs in activation checks.
+- **Coordination Challenges:** Decoupling might lead to inconsistent feature enablement
```

Comment:
I feel strongly that once we decide what goes into a hardfork, we disable the ability to modify the feature flags that go within it.

Reply:
Two thoughts here:
1. If features are defined in SCR, it makes it difficult or at least a very explicit change in client software to enable or disable features out-of-band (for the intended hardfork)
2. In contrast, if features are only handled in development, changes to these features in forked clients are somewhat more expected.